src/test-macro/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+#![warn(clippy::print_literal)]
+#![warn(clippy::same_item_push)]
+#![warn(clippy::useless_format)]
+#![warn(clippy::write_literal)]
+#![warn(clippy::redundant_closure)]
+#![warn(clippy::redundant_closure_call)]
+#![warn(clippy::unnecessary_lazy_evaluations)]
+#![warn(clippy::partialeq_ne_impl)]
+#![warn(clippy::redundant_field_names)]
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
+#![warn(clippy::unused_async)]
+#![warn(clippy::disallowed_methods)]
+#![warn(clippy::disallowed_macros)]
+#![warn(clippy::disallowed_types)]
+#![warn(clippy::from_over_into)]
+// END LINT CONFIG
+
+//! test macro with auto-initialized logging
+
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as Tokens;
+use quote::quote;
+use syn::parse_macro_input;
+use syn::parse_quote;
+use syn::AttributeArgs;
+use syn::ItemFn;
+use syn::Meta;
+use syn::NestedMeta;
+use syn::ReturnType;
+
+/// Based on <https://github.com/d-e-s-o/test-log>
+/// Copyright (C) 2019-2022 Daniel Mueller <deso@posteo.net>
+/// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+#[proc_macro_attribute]
+pub fn test(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(attr as AttributeArgs);
+    let input = parse_macro_input!(item as ItemFn);
+
+    let inner_test = match args.as_slice() {
+        [] => parse_quote! { ::core::prelude::v1::test },
+        [NestedMeta::Meta(Meta::Path(path))] => quote! { #path },
+        [NestedMeta::Meta(Meta::List(list))] => quote! { #list },
+        _ => panic!("unsupported attributes supplied: {:?}", args),
+    };
+
+    expand_wrapper(&inner_test, &input)
+}
+
+fn expand_logging_init() -> Tokens {
+    let crate_name = std::env::var("CARGO_PKG_NAME").unwrap();
+    if crate_name == "mz-ore" {
+        quote! {
+          {
+            use crate::test;
+            let _ = test::init_logging();
+          }
+        }
+    } else {
+        quote! {
+          {
+            let _ = ::mz_ore::test::init_logging();
+          }
+        }
+    }
+}
+
+/// Emit code for a wrapper function around a test function.
+fn expand_wrapper(inner_test: &Tokens, wrappee: &ItemFn) -> TokenStream {
+    let attrs = &wrappee.attrs;
+    let async_ = &wrappee.sig.asyncness;
+    let await_ = if async_.is_some() {
+        quote! {.await}
+    } else {
+        quote! {}
+    };
+    let body = &wrappee.block;
+    let test_name = &wrappee.sig.ident;
+
+    // Note that Rust does not allow us to have a test function with
+    // #[should_panic] that has a non-unit return value.
+    let ret = match &wrappee.sig.output {
+        ReturnType::Default => quote! {},
+        ReturnType::Type(_, type_) => quote! {-> #type_},
+    };
+
+    let logging_init = expand_logging_init();
+
+    let result = quote! {
+      #[#inner_test]
+      #(#attrs)*
+      #async_ fn #test_name() #ret {
+        #async_ fn test_impl() #ret {
+          #body
+        }
+
+        #logging_init
```

Comment:
If we later have other things we want to run in each unit test, during setup or at the end, we can add them here in the future.